pkg/controllers/vdb/shutdownspec_reconciler.go

```diff
@@ -0,0 +1,94 @@
+ You may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package vdb
+
+import (
+	"context"
+
+	v1 "github.com/vertica/vertica-kubernetes/api/v1"
+	"github.com/vertica/vertica-kubernetes/pkg/controllers"
+	vmeta "github.com/vertica/vertica-kubernetes/pkg/meta"
+	"github.com/vertica/vertica-kubernetes/pkg/vk8s"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+// ShutdownSpecReconciler will make sure that subclusters, part of a sandbox,
+// that needs to be shutdown/restarted, have the correct shutdown value.
+type ShutdownSpecReconciler struct {
+	VRec   *VerticaDBReconciler
+	Vdb    *v1.VerticaDB
+	PFacts *PodFacts
+}
+
+func MakeShutdownSpecReconciler(r *VerticaDBReconciler,
+	vdb *v1.VerticaDB, pfacts *PodFacts) controllers.ReconcileActor {
+	return &ShutdownSpecReconciler{
+		VRec:   r,
+		Vdb:    vdb,
+		PFacts: pfacts,
+	}
+}
+
+func (r *ShutdownSpecReconciler) Reconcile(ctx context.Context, _ *ctrl.Request) (ctrl.Result, error) {
+	// no-op as there is no sandbox
+	if len(r.Vdb.Spec.Sandboxes) == 0 {
+		return ctrl.Result{}, nil
+	}
+	return ctrl.Result{}, r.updateSubclustersShutdownState(ctx)
+}
+
+// updateSubclustersShutdownState updates the sandbox's subclusters shutdown state after
+// the sandbox has been restarted.
+func (r *ShutdownSpecReconciler) updateSubclustersShutdownState(ctx context.Context) error {
+	_, err := vk8s.UpdateVDBWithRetry(ctx, r.VRec, r.Vdb, r.updateSubclustersShutdownStateCallback)
+	return err
+}
+
+func (r *ShutdownSpecReconciler) updateSubclustersShutdownStateCallback() (bool, error) {
+	needUpdate := false
+	scMap := r.Vdb.GenSubclusterMap()
+	for i := range r.Vdb.Spec.Sandboxes {
+		sb := &r.Vdb.Spec.Sandboxes[i]
+		for j := range sb.Subclusters {
+			sc := scMap[sb.Subclusters[j].Name]
+			// Proceed only if subcluster's shutdown is not equal
+			// to sandbox's shutdown
+			if sb.Shutdown == sc.Shutdown {
+				continue
+			}
+			if sb.Shutdown {
+				if sc.Annotations == nil {
+					sc.Annotations = make(map[string]string, 1)
+				}
+				// Add a label that indicate the shutdown/restart is controlled
+				// by the sandbox as opposed to the subcluster. It helps
+				// differentiate this case from when the user is explicitly
+				// changing the subcluster's shutdown field.
+				sc.Annotations[vmeta.ShutdownDrivenBySandbox] = "true"
+			} else {
+				// If the shutdown/restart is not controlled by the sandbox,
+				// we skip to the next subcluster.
+				if !vmeta.GetShutdownDrivenBySandbox(sc.Annotations) {
+					continue
+				}
+				delete(sc.Annotations, vmeta.ShutdownDrivenBySandbox)
```

Comment:
We should add "continue" after this line. When sb.Shutdown == false, we don't want to execute line 89 since single subcluster can be shut down.

Reply:
Line 84 takes care of that. The "shutdown-driven-by-sandbox" is set only when shutdown is done through spec.sandboxes[]. shutdown(total sandbox shutdown). This reconciler will not touch subclusters whose shutdown is controlled by the user(user explicitly changed the subcluster's shutdown field).
There will be webhook rules to make sure users cannot change a  sandbox subcluster's shutdown field if the shutdown is controlled by the sandbox(total sandbox shutdown). This feature can keep users clusters down if they are not careful that's why I want to control its use.

Reply:
That makes sense with webhook rules.